oauth/authprovider.go

```diff
@@ -0,0 +1,122 @@
+package oauth
+
+import (
+	"errors"
+	"sync"
+)
+
+type Token struct {
+	AccessToken  string `json:"access_token"`
+	RefreshToken string `json:"refresh_token"`
+}
+
+type AuthenticationProvider interface {
+	GenerateAuthCode(clientId string, agentUserId string) (string, error)
+	GenerateToken(clientId string, clientSecret string, authCode string) (Token, error)
+	GetAgentUserIdForToken(token string) (agentUserId string, refresh bool, err error)
+}
+
+// a very simple authentication provider
+
+type SimpleAuthenticationProvider struct {
```

Comment:
This struct needs to be re-worked to save its data on disk to survive reboots of the program,

Reply:
This struct should also probably be moved into its own package

Reply:
Or better yet, the concept of "Registered agents" should be abstracted away from the oauth authcodes/tokens etc

Reply:
Maybe just serialize the struct as json and write it to disk?

Reply:
this was my plan :P I might re-work the struct some before doing so

Reply:
See #10 for this